### PR TITLE
Instrument `_check_sigs_and_hash_and_fetch` to trace time spent in child concurrent calls

### PR DESCRIPTION
**Pull request by MadLittleMods**
Instrument `_check_sigs_and_hash_and_fetch` to trace time spent in child concurrent calls because I've see `_check_sigs_and_hash_and_fetch` take [10.41s to process 100 events](https://github.com/matrix-org/synapse/issues/13587).

Fix https://github.com/matrix-org/synapse/issues/13587

Part of https://github.com/matrix-org/synapse/issues/13356

Interestingly, each `pdu` takes incrementally longer as you go,
Jaeger trace with the new instrumentation

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
  - Feel free to credit yourself, by adding a sentence "Contributed by @github_username." or "Contributed by [Your Name]." to the end of the entry.
* [ ] ~~Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)~~
* [x] [Code style](https://matrix-org.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

### CODE REVIEW

File: synapse/federation/federation_base.py [hunk 1]
```diff
@@ -97,17 +99,36 @@ async def _check_sigs_and_hash(
...
 
         if spam_check != self.spam_checker.NOT_SPAM:
             logger.warning("Event contains spam, soft-failing %s", pdu.event_id)
+            log_kv(
+                {
+                    "message": "Event contains spam, redacting (to save disk space) "
+                    "as well as soft-failing (to stop using the event in prev_events)",
+                    "event_id": pdu.event_id,
+                }
+            )
```
Comment (MadLittleMods) [inline synapse/federation/federation_base.py:131]: This part is very interesting. I wonder if we're seeing some of the backfilled messages as spam, soft-failing them, which causes the `Unable to get missing prev_event $luA4l7QHhf_jadH3mI-AyFqho0U2Q-IXXUbGSMq6h6M` error when trying to figure out `_get_state_ids_after_missing_prev_event`.

We should definitely skip trying to backfill events we see as spam over and over.

File: synapse/federation/federation_client.py [hunk 2]
```diff
@@ -643,6 +649,14 @@ async def _check_sigs_and_hash_and_fetch_one(
                 pdu.event_id,
                 e,
             )
+            log_kv(
+                {
+                    "message": "Signature on retrieved event was invalid. "
+                    "Checking local store/orgin server",
```
Comment (MadLittleMods) [inline synapse/federation/federation_client.py:655]: Why do we `get_pdu` from a remote origin homeserver again?

The events we pass to `_check_sigs_and_hash_and_fetch` were already fetched and that's why we're checking them here.

File: synapse/util/ratelimitutils.py [hunk 3]
```diff
@@ -168,6 +168,11 @@ def should_reject(self) -> bool:
         (either sleeping or in the ready queue).
         """
         queue_size = len(self.ready_request_queue) + len(self.sleeping_requests)
+        logger.info(
+            "should_reject queue_size=%s self.reject_limit=%s",
+            queue_size,
+            self.reject_limit,
+        )
```
Comment (MadLittleMods) [inline synapse/util/ratelimitutils.py:175]: TODO: Remove this debug log for another problem

File: synapse/federation/federation_client.py (hunk 2 above)
Comment (MadLittleMods) [reply synapse/federation/federation_client.py:655]: From the [comment doc](https://github.com/matrix-org/synapse/blob/3dd175b628bab5638165f20de9eade36a4e88147/synapse/federation/federation_client.py#L570-L571), it looks like we ask the original server that sent the message which could be different than how we encountered it initially.

> request it from the sender's server (if that is different from `origin`)

File: synapse/federation/federation_client.py [hunk 4]
```diff
@@ -643,12 +649,23 @@ async def _check_sigs_and_hash_and_fetch_one(
...
```
Comment (reivilibre) [inline synapse/federation/federation_client.py:655]:
```suggestion
                    "Checking local store/origin server",
```

**Review by reivilibre — APPROVED** [review]
Look like reasonable debug points to me

**Comment by MadLittleMods** [comment]
Thanks for the review @reivilibre 🐭

**Pull request merged**